content/blog/2022/2022-09-02-junit5.md

```diff
@@ -0,0 +1,601 @@
+---
+title: "Complete Guide to JUnit 5"
+categories: ["Java"]
+date: 2022-09-02 00:00:00 +1100
+authors: ["cercenazi"]
+description: "Complete Guide to JUnit 5."
+image: images/stock/0125-junit5-1200x628.jpg
+url: junit5
+---
+In software development, unit testing means writing tests to verify that our code is working as logical units. Different people tend to think of a "logical unit" in different ways. We will define it as a unit of behavior that tests not a method or a class on it own but as the collective behavior of many methods that does a useful business work.
+Unit testing allows us to catch bugs early on in the development phase plus allowing us to refactor and change our code knowing that if we mess up something the tests will light red and notify us.
+
+{{% github "https://github.com/thombergs/code-examples/tree/master/core-java/junit5" %}}
+
+## What is JUnit and JUnit5
+JUnit is an Open Source testing framework for Java which relies heavily on annotations to run and manage our tests. It allows us to write our tests in logically separated suites which are executed in parallel fast runs.
+
+### JUnit5
+It was developed to leverage the new and powerful advances in the Java language introduced in Java8 and beyond. It embraces the functional and declarative style of programming which is more readable and easier to write.
+JUnit 5 can use more than one extension at a time, which was not possible in earlier versions where only one runner could be used at a time. This allows us to combine Spring extensions with other extensions including custom ones that we write.
+
+## Basic components of the JUnit5 project
+
+It composes of three main parts
+- JUnit Platform
+  - Serves as a foundation for launching testing frameworks on the JVM.
+  - Defines the TestEngine API for developing a testing framework that runs on the platform.
+
+- JUnit Jupiter
+  - Offers us the annotations and the extensions to write the tests.
+
+- JUnit Vintage
+  - Offers us the ability to run JUnit4 and JUnit3 tests on the platform.
+
+
+## Annotations
+JUnit annotations are offered to us by the Jupiter model. They make our tests much more readable and easy to write. Let's take a look at the most important ones.
+
+Before we start with the code we need to add the `junit-jupiter-api` dependency and the `maven-surefire-plugin` plugin to be able to run the tests using Maven commands.
```

Comment:
Isn't the surefire plugin installed by default in Maven?

Reply:
Yes it seems so :)